From 90c037be8ee0ec9c16d035fc067a8aa9554ea2c7 Mon Sep 17 00:00:00 2001
From: Michael Forney <mforney@mforney.org>
Date: Wed, 13 Feb 2019 16:39:52 -0800
Subject: [PATCH] amd64: Make floating constants their own temporaries

Otherwise, we may attempt a memory-to-memory mov if trying to save the
constant to a slot (for example a spilled phi with a constant float
argument).
---
 amd64/isel.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/amd64/isel.c b/amd64/isel.c
index 6aea850..babcd65 100644
--- a/amd64/isel.c
+++ b/amd64/isel.c
@@ -84,6 +84,9 @@ fixarg(Ref *r, int k, int op, Fn *fn)
 		sprintf(buf, "fp%d", n);
 		a.offset.label = intern(buf);
 		fn->mem[fn->nmem-1] = a;
+		r0 = r1;
+		r1 = newtmp("isel", k, fn);
+		emit(Ocopy, k, r1, r0, R);
 	}
 	else if (!cpy && k == Kl && noimm(r0, fn)) {
 		/* load constants that do not fit in
-- 
2.20.1

